pkg/storage/unified/resource/server.go

```diff
@@ -581,21 +622,37 @@ func (s *server) Watch(req *WatchRequest, srv ResourceStore_WatchServer) error {
 			}
 
 			if event.ResourceVersion > since && matchesQueryKey(req.Options.Key, event.Key) {
-				// Currently sending *every* event
-				// if req.Options.Labels != nil {
-				// 	// match *either* the old or new object
-				// }
-				// TODO: return values that match either the old or the new
-
-				if err := srv.Send(&WatchEvent{
+				value := event.Value
+				// remove the delete marker stored in the value for deleted objects
+				if event.Type == WatchEvent_DELETED {
+					value = []byte{}
+				}
+				resp := &WatchEvent{
 					Timestamp: event.Timestamp,
 					Type:      event.Type,
 					Resource: &WatchEvent_Resource{
-						Value:   event.Value,
+						Value:   value,
 						Version: event.ResourceVersion,
 					},
-					// TODO... previous???
-				}); err != nil {
+				}
+				if event.PreviousRV > 0 {
+					prevObj, err := s.Read(ctx, &ReadRequest{Key: event.Key, ResourceVersion: event.PreviousRV})
+					if err != nil {
+						// TODO: Are we expecting to discard the event if the previous object is not found? Or should we send the event with the current object only?
+						// Failing to send the previous object means consumers might see get the event when using precondition checks.
+						// For now, we will sent the event with the current object only and log the error.
```

Comment:
I think the Kubernetes API in general defaults to use `synthetic "Added" events` (in their docs own words) whenever historic data isn't available for any reason.

See this section, for example: https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-watch

The other alternative is failing altogether with a 410 Gone: https://kubernetes.io/docs/reference/using-api/api-concepts/#410-gone-responses

Reply:
This scenario doesn't really exist on k8s side. On watch, etcd always has access to the previous event (by design, [here](https://github.com/kubernetes/kubernetes/blob/37f733a657ef71d66177d00f9b7d47ec507dedd3/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L468), the previous value is part of the etcd watch event).

In our implementation, this shouldn't really happen either 🤔 . So, returning an error is probably ok ?

Reply:
Ok. Seems that the client is already handling this case. What is does depends on the actual event.  Best leave it like this.
https://github.com/grafana/grafana/blob/5b551fd7f9c0f97b28dc39781593788db46fea21/pkg/storage/unified/apistore/stream.go#L164-L170